File: 2023-2024/sourcecode/mod2/Lab7/MVVMBasicsJava/app/src/main/java/com/example/mvvmbasicsjava/ui/fragment/UsersFragment.java
package com.example.mvvmbasicsjava.ui.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.mvvmbasicsjava.databinding.FragmentUsersBinding;
import com.example.mvvmbasicsjava.model.User;
import com.example.mvvmbasicsjava.viewmodel.UserViewModel;

import java.util.ArrayList;

public class UsersFragment extends Fragment {

    private FragmentUsersBinding binding;

    private UserViewModel viewModel;
    private UserAdapter userAdapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentUsersBinding.inflate(inflater, container, false);

        viewModel = new ViewModelProvider(this).get(UserViewModel.class);
        userAdapter = new UserAdapter(new UserComparator());

        viewModel.getUsersList().observe(getViewLifecycleOwner(), users -> userAdapter.submitList(new ArrayList<>(users)));

        binding.rvList.setAdapter(userAdapter);
        binding.rvList.setLayoutManager(new LinearLayoutManager(requireContext()));

        binding.addButton.setOnClickListener(v -> onAddUser());
        binding.resetButton.setOnClickListener(v -> onResetUsers());
        binding.clearButton.setOnClickListener(v -> onClearUsers());

        return binding.getRoot();
    }

    private void onAddUser() {
        if (binding.firstNameEditText.getText() != null && binding.lastNameEditText.getText() != null) {
            String firstName = binding.firstNameEditText.getText().toString().trim();
            String lastName = binding.lastNameEditText.getText().toString().trim();

            if (!firstName.isEmpty() && !lastName.isEmpty()) {
                User newUser = new User(firstName, lastName);
                viewModel.addUser(newUser);

                binding.firstNameEditText.getText().clear();
                binding.lastNameEditText.getText().clear();
            }
        }
    }

    private void onResetUsers() {
        viewModel.reinitialize();
    }

    private void onClearUsers() {
        viewModel.clear();
    }
}

File: 2023-2024/sourcecode/mod2/Lab7/MVVMBasicsKotlin/app/src/main/java/com/example/mvvmbasicskotlin/ui/fragment/UsersFragment.kt
package com.example.mvvmbasicskotlin.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.mvvmbasicskotlin.databinding.FragmentUsersBinding
import com.example.mvvmbasicskotlin.model.User
import com.example.mvvmbasicskotlin.viewmodel.UserViewModel

class UsersFragment : Fragment() {

    private lateinit var binding: FragmentUsersBinding

    private val viewModel: UserViewModel by viewModels()
    private val userAdapter by lazy { UserAdapter(UserComparator())}

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentUsersBinding.inflate(inflater)


        viewModel.usersList.observe(viewLifecycleOwner) { users ->
            userAdapter.submitList(users.toMutableList())
        }

        binding.rvList.apply{
            adapter = userAdapter
            layoutManager = LinearLayoutManager(requireContext())
        }

        binding.addButton.setOnClickListener { onAddUser() }
        binding.resetButton.setOnClickListener { onResetUsers() }
        binding.clearButton.setOnClickListener { onClearUsers() }

        return binding.root
    }

    private fun onAddUser() {
        val firstName = binding.firstNameEditText.text.toString().trim()
        val lastName = binding.lastNameEditText.text.toString().trim()

        if (firstName.isNotBlank() && lastName.isNotBlank()) {
            val newUser = User(firstName, lastName)
            viewModel.addUser(newUser)

            binding.firstNameEditText.text?.clear()
            binding.lastNameEditText.text?.clear()
        }
    }

    private fun onResetUsers() {
        viewModel.reinitialize()
    }

    private fun onClearUsers() {
        viewModel.clear()
    }
}